### PR TITLE
Create ansible setup for Program Manager to be in sandbox build

### PR DESCRIPTION
**Pull request by schenedx**
Configuration Pull Request
Add Program Manager MFE
---

Make sure that the following steps are done before merging:

  - [ ] A DevOps team member has approved the PR if it is code shared across multiple services and you don't own all of the services.
  - [x] If you are making a complicated change, have you performed the proper testing specified on the [Ops Ansible Testing Checklist](https://openedx.atlassian.net/wiki/display/EdxOps/Ops+Ansible+Testing+Checklist)?  Adding a new variable does not require the full list (although testing on a sandbox is a great idea to ensure it links with your downstream code changes).

### CODE REVIEW

File: playbooks/roles/program_manager/defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+PROGRAM_MANAGER_REGISTRAR_API_BASE_URL: !!null
+PROGRAM_MANAGER_DISCOVERY_BASE_URL: !!null
+PROGRAM_MANAGER_NGINX_PORT: 80
+PROGRAM_MANAGER_SSL_NGINX_PORT: 443
+
+program_manager_home: '{{ COMMON_APP_DIR }}/{{ program_manager_service_name }}'
+NVM_DIR: '{{ program_manager_home }}'
+program_manager_user: 'root'
+program_manager_git_identity: 'none'
+edx_django_service_use_python3: false
```
Comment (iloveagent57) [inline playbooks/roles/program_manager/defaults/main.yml]: does this need to be here?

File: playbooks/roles/program_manager/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+program_manager_dist_dir: "{{ program_manager_code_dir }}/dist"
+program_manager_env_vars:
+  PATH: "{{ program_manager_nodeenv_bin }}:{{ program_manager_venv_dir }}/bin:{{ ansible_env.PATH }}"
+  NODE_ENV: "production"
+  ACTIVE_ENV: "production"
+  BASE_URL: "{{ PROGRAM_MANAGER_URL_ROOT }}"
+  LMS_BASE_URL: "{{ PROGRAM_MANAGER_LMS_BASE_URL }}"
+  REGISTRAR_API_BASE_URL: "{{ PROGRAM_MANAGER_REGISTRAR_API_BASE_URL }}"
+  DISCOVERY_BASE_URL: "{{ PROGRAM_MANAGER_DISCOVERY_BASE_URL }}"
+  CREDENTIALS_API_BASE_URL: 'http://credentials-{{ COMMON_LMS_BASE_URL }}/api'
```
Comment (iloveagent57) [inline playbooks/roles/program_manager/defaults/main.yml]: Is there a `CREDENTIALS_BASE_URL` we can use instead?  You probably want https:// in stage and prod, right?

File: playbooks/roles/program_manager/defaults/main.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  REFRESH_ACCESS_TOKEN_ENDPOINT: '{{ COMMON_LMS_BASE_URL }}/login_refresh'
+  ACCESS_TOKEN_COOKIE_NAME: 'edx-jwt-cookie-header-payload'
+  USER_INFO_COOKIE_NAME: 'edx-user-info'
+  MARKETING_SITE_BASE_URL: 'https://stage.edx.org'
+  SUPPORT_URL: 'https://stage.edx.org/support'
+  CONTACT_URL: 'https://stage.edx.org/contact'
+  OPEN_SOURCE_URL: 'https://stage.edx.org/openedx'
+  TERMS_OF_SERVICE_URL: 'https://stage.edx.org/terms-of-service'
+  PRIVACY_POLICY_URL: 'https://stage.edx.org/privacy-policy'
+  FACEBOOK_URL: 'https://www.facebook.com'
```
Comment (iloveagent57) [inline playbooks/roles/program_manager/defaults/main.yml:53]: do we need all of these?

File: playbooks/roles/program_manager/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+- name: Checkout program-manager repo into {{ program_manager_code_dir }}
+  git:
+    dest: "{{ program_manager_code_dir }}"
+    repo: "{{ program_manager_repo }}"
+    version: "{{ program_manager_version }}"
+    accept_hostkey: yes
+  become_user: "{{ program_manager_user }}"
+  register: program_manager_checkout
+
+- name: build virtualenv with python2.7
```
Comment (iloveagent57) [inline playbooks/roles/program_manager/tasks/main.yml]: why do we need this?

File: playbooks/roles/program_manager/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  command: "virtualenv --python=python2.7 {{ program_manager_venv_dir }}"
+  args:
+    creates: "{{ program_manager_venv_dir }}/bin/pip"
+  become_user: "{{ program_manager_user }}"
+  when: not edx_django_service_use_python3
+  tags:
+    - install
+    - install:system-requirements
+
+- name: Pin pip to a specific version.
```
Comment (iloveagent57) [inline playbooks/roles/program_manager/tasks/main.yml]: why do we need this?

File: playbooks/roles/program_manager/defaults/main.yml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+program_manager_home: '{{ COMMON_APP_DIR }}/{{ program_manager_service_name }}'
+NVM_DIR: '{{ program_manager_home }}'
+program_manager_user: 'root'
+program_manager_git_identity: 'none'
+edx_django_service_use_python3: false
+program_manager_repo: 'https://github.com/edx/frontend-app-program-manager.git'
+program_manager_version: 'master'
+program_manager_service_name: 'program-manager'
+program_manager_node_version: '10.16.0'
+program_manager_venv_dir: '{{ program_manager_home }}/venvs/{{ program_manager_service_name }}'
```
Comment (iloveagent57) [inline playbooks/roles/program_manager/defaults/main.yml]: we shouldn't need a python virtualenv for a node app, right?

File: playbooks/roles/program_manager/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  command: "{{ program_manager_venv_dir }}/bin/pip install pip=={{ common_pip_version }}"
+  become_user: "{{ program_manager_user }}"
+  tags:
+    - install
+    - install:system-requirements
+
+# NOTE (CCB): Ideally we should use the pip Ansible command,
+# but that doesn't seem to work with the Python 3.x virtualenv.
+- name: install nodenv
+  command: pip install nodeenv
```
Comment (iloveagent57) [inline playbooks/roles/program_manager/tasks/main.yml]: wait, we need a python virtualenv just to install node?  that seems counterintuitive

**Review by iloveagent57 — COMMENTED** [review]
I have some questions about cruft, but if we really need a python virtualenv to install node, I guess that's fine.

File: playbooks/roles/program_manager/defaults/main.yml (hunk 3 above)
Comment (schenedx) [reply playbooks/roles/program_manager/defaults/main.yml:53]: These are for the MFE footers. Might as well specify them

File: playbooks/roles/program_manager/tasks/main.yml (hunk 4 above)
Comment (schenedx) [reply playbooks/roles/program_manager/tasks/main.yml]: I actually don't know. I can test to remove and see if that make a difference

File: playbooks/roles/program_manager/defaults/main.yml (hunk 6 above)
Comment (schenedx) [reply playbooks/roles/program_manager/defaults/main.yml]: I hope so. I can test to remove

**Comment by schenedx** [comment]
So, we use `nodeenv` which is a python library to install the nodejs version we want.
If we don't use `nodeenv`, we would have to write ansible tasks to download the node version from internet and install nodejs by doing `apt-get install` command. I rather have `nodeenv` take care of that for me. Let me know if you disagree

File: playbooks/roles/program_manager/defaults/main.yml (hunk 2 above)
Comment (schenedx) [reply playbooks/roles/program_manager/defaults/main.yml]: This is not used in Program-manager. I am removing it.

File: playbooks/roles/program_manager/defaults/main.yml (hunk 1 above)
Comment (schenedx) [reply playbooks/roles/program_manager/defaults/main.yml]: This is needed. Apparently, if I remove it, the error message is 
```
fatal: [localhost]: FAILED! => {"msg": "The conditional check 'not
 edx_django_service_use_python3' failed. The error was: error while evaluating conditional
 (not edx_django_service_use_python3): 'edx_django_service_use_python3' is 
undefined\n\nThe error appears to have been in '/edx/app/edx_ansible/edx_ansible/playbooks/roles/program_manager/tasks/main.yml': line
 27, column 3, but may\nbe elsewhere in the file depending on the exact syntax 
problem.\n\nThe offending line appears to be:\n\n\n- name: build virtualenv with 
python2.7\n  ^ here\n"}
```

File: playbooks/roles/program_manager/defaults/main.yml (hunk 3 above)
Comment (iloveagent57) [reply playbooks/roles/program_manager/defaults/main.yml:53]: Can't the footer component specify these, then?  Things like facebook and twitter URLs are really, really unlikely to change and could basically be hard-coded there.  Not a blocker, but it'd be cool to see them removed from here in the near future.

File: playbooks/roles/program_manager/defaults/main.yml (hunk 1 above)
Comment (iloveagent57) [reply playbooks/roles/program_manager/defaults/main.yml]: Alright, fair enough, but for the record, I think this is janky.

**Review by iloveagent57 — APPROVED** [review]

File: playbooks/roles/program_manager/tasks/main.yml (hunk 4 above)
Comment (fredsmith) [reply playbooks/roles/program_manager/tasks/main.yml]: if you remove all these tasks, you can remove the unnecessary python variables above.

File: playbooks/roles/program_manager/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    accept_hostkey: yes
+  become_user: "{{ program_manager_user }}"
+  register: program_manager_checkout
+
+- name: build virtualenv with python2.7
+  command: "virtualenv --python=python2.7 {{ program_manager_venv_dir }}"
+  args:
+    creates: "{{ program_manager_venv_dir }}/bin/pip"
+  become_user: "{{ program_manager_user }}"
+  when: not edx_django_service_use_python3
```
Comment (fredsmith) [inline playbooks/roles/program_manager/tasks/main.yml]: this when: clause is why you're getting the error that @iloveagent57 says, and I agree, is janky

File: playbooks/roles/program_manager/tasks/main.yml (hunk 7 above)
Comment (fredsmith) [reply playbooks/roles/program_manager/tasks/main.yml]: nodeenv is in apt.  don't use pip.

File: playbooks/roles/program_manager/tasks/main.yml (hunk 5 above)
Comment (fredsmith) [reply playbooks/roles/program_manager/tasks/main.yml]: you don't, kill it.

File: playbooks/roles/program_manager/defaults/main.yml (hunk 3 above)
Comment (fredsmith) [reply playbooks/roles/program_manager/defaults/main.yml:53]: I agree, this seems unnecessary to pass into every program execution in the env.

**Review by fredsmith — CHANGES_REQUESTED** [review]
no need to do all the python stuff just to get nodeenv, use the one in apt.

**Comment by schenedx** [comment]
@fredsmith The nodejs in apt-get, or dpkg is version 4.2.6. It's not the version 10.16.0 we want for this application. Can you tell me what you mean by `nodeenv is in apt`?

**Comment by fredsmith** [comment]
> @fredsmith The nodejs in apt-get, or dpkg is version 4.2.6. It's not the version 10.16.0 we want for this application. Can you tell me what you mean by `nodeenv is in apt`?

I mean do this:

```
$ apt install nodeenv
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following package was automatically installed and is no longer required:
  libllvm7
Use 'sudo apt autoremove' to remove it.
The following NEW packages will be installed:
  nodeenv
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 18.1 kB of archives.
After this operation, 81.9 kB of additional disk space will be used.
Get:1 http://us.archive.ubuntu.com/ubuntu bionic/universe amd64 nodeenv all 0.13.4-1 [18.1 kB]
Fetched 18.1 kB in 0s (152 kB/s)   
Selecting previously unselected package nodeenv.
(Reading database ... 180118 files and directories currently installed.)
Preparing to unpack .../nodeenv_0.13.4-1_all.deb ...
Unpacking nodeenv (0.13.4-1) ...
Setting up nodeenv (0.13.4-1) ...
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...

$ nodeenv test --node=10.16.0 --prebuilt --force
 * Install node (10.16.0)... done.

$ . test/bin/activate 

$ node --version
v10.16.0
```

Use this module:  https://docs.ansible.com/ansible/latest/modules/apt_module.html to apt install the nodeenv package instead of installing pip and making a virtualenv and pip installing nodeenv just to install node

**Comment by schenedx** [comment]
@fredsmith Make sense now. I'll try to update my PR shortly

**Comment by schenedx** [comment]
@fredsmith Changes made. Could you review again? I manually tested the newest changes here and it's all working.

**Review by fredsmith — COMMENTED** [review]
Looks good :shipit:

**Review by fredsmith — APPROVED** [review]

**Pull request merged**